engines/e_pkcs11.c

```diff
@@ -0,0 +1,490 @@
+                  PKCS11_R_LIBRARY_PATH_NOT_FOUND);
+        return CKR_GENERAL_ERROR;
+    }
+
+    pFunc = (CK_RV (*)()) DSO_bind_func(pkcs11_dso, "C_GetFunctionList");
+
+    if (pFunc == NULL) {
+        PKCS11_trace("C_GetFunctionList() not found in module %s\n",
+                     library_path);
+        PKCS11err(PKCS11_F_PKCS11_LOAD_FUNCTIONS,
+                  PKCS11_R_GETFUNCTIONLIST_NOT_FOUND);
+        return CKR_FUNCTION_NOT_SUPPORTED;
+    }
+
+    rv = pFunc(&pkcs11_funcs);
+
+    return rv;
+}
+
+/**
+ * Initialize the PKCS#11 library.
+ * This loads the function list and initializes PKCS#11.
+ * @param library_path
+ * @return
+ */
+CK_RV pkcs11_initialize(char *library_path)
+{
+    CK_RV rv;
+
+    if (!library_path) {
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    rv = pkcs11_load_functions(library_path);
+    if (rv != CKR_OK) {
+        PKCS11_trace("Getting PKCS11 function list failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_INITIALIZE,
+                  PKCS11_R_GETTING_FUNCTION_LIST_FAILED);
+        return rv;
+    }
+
+    CK_C_INITIALIZE_ARGS args;
+    memset(&args, 0, sizeof(args));
+    args.flags = CKF_OS_LOCKING_OK;
+    rv = pkcs11_funcs->C_Initialize(&args);
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_Initialize failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_INITIALIZE, PKCS11_R_INITIALIZE_FAILED);
+        return rv;
+    }
+
+    return CKR_OK;
+}
+
+static void pkcs11_finalize()
+{
+    pkcs11_funcs->C_Finalize(NULL);
+}
+
+static CK_SLOT_ID pkcs11_get_slot()
+{
+    CK_RV rv;
+    CK_SLOT_ID slotId;
+    CK_ULONG slotCount = 10;
+    CK_SLOT_ID *slotIds = OPENSSL_malloc(sizeof(CK_SLOT_ID) * slotCount);
+    rv = pkcs11_funcs->C_GetSlotList(CK_TRUE, slotIds, &slotCount);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetSlotList failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, PKCS11_R_GET_SLOTLIST_FAILED);
+        goto err;
+    }
+
+    if (slotCount < 1) {
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, PKCS11_R_SLOT_NOT_FOUND);
+        goto err;
+    }
+
+    slotId = slotIds[0];
+    OPENSSL_free(slotIds);
+    return slotId;
+
+ err:
+    free(slotIds);
+    return 0;
+}
+
+static CK_SESSION_HANDLE pkcs11_start_session(CK_SLOT_ID slotId)
+{
+    CK_RV rv;
+    CK_SESSION_HANDLE session;
+    rv = pkcs11_funcs->C_OpenSession(slotId, CKF_SERIAL_SESSION, NULL,
+                                     NULL, &session);
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_OpenSession failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_START_SESSION,
+                  PKCS11_R_OPEN_SESSION_ERROR);
+        goto err;
+    }
+    return session;
+
+ err:
+    return 0;
+}
+
+static int pkcs11_login(CK_SESSION_HANDLE session, CK_BYTE *pin, CK_USER_TYPE userType)
+{
+    CK_RV rv;
+    if (pin) {
+        rv = pkcs11_funcs->C_Login(session, userType, pin,
+                                   strlen((char *)pin));
```

Comment:
Note that PINs can be binary - you are assuming a string value here - which is true in this context - but in general it isn't. 
So just a comment on the internal function not supporting binary pins.